source/common/http/filter/lua/lua_filter.cc

```diff
@@ -0,0 +1,490 @@
+#include "common/http/filter/lua/lua_filter.h"
+
+#include "envoy/http/codes.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+#include "common/common/enum_to_int.h"
+#include "common/http/message_impl.h"
+
+namespace Envoy {
+namespace Http {
+namespace Filter {
+namespace Lua {
+
+StreamHandleWrapper::StreamHandleWrapper(Envoy::Lua::CoroutinePtr&& coroutine, HeaderMap& headers,
+                                         bool end_stream, Filter& filter,
+                                         FilterCallbacks& callbacks)
+    : coroutine_(std::move(coroutine)), headers_(headers), end_stream_(end_stream), filter_(filter),
+      callbacks_(callbacks), yield_callback_([this]() {
+        if (state_ == State::Running) {
+          throw Envoy::Lua::LuaException("script performed an unexpected yield");
+        }
+      }) {}
+
+FilterHeadersStatus StreamHandleWrapper::start(int function_ref) {
+  // We are on the top of the stack.
+  coroutine_->start(function_ref, 1, yield_callback_);
+  FilterHeadersStatus status =
+      (state_ == State::WaitForBody || state_ == State::HttpCall || state_ == State::Responded)
+          ? FilterHeadersStatus::StopIteration
+          : FilterHeadersStatus::Continue;
+
+  if (status == FilterHeadersStatus::Continue) {
+    headers_continued_ = true;
+  }
+
+  return status;
+}
+
+FilterDataStatus StreamHandleWrapper::onData(Buffer::Instance& data, bool end_stream) {
+  ASSERT(!end_stream_);
+  end_stream_ = end_stream;
+  saw_body_ = true;
+
+  if (state_ == State::WaitForBodyChunk) {
+    ENVOY_LOG(debug, "resuming for next body chunk");
+    Envoy::Lua::LuaDeathRef<Envoy::Lua::BufferWrapper> wrapper(
+        Envoy::Lua::BufferWrapper::create(coroutine_->luaState(), data), true);
+    state_ = State::Running;
+    coroutine_->resume(1, yield_callback_);
+  } else if (state_ == State::WaitForBody && end_stream_) {
+    ENVOY_LOG(debug, "resuming body due to end stream");
+    callbacks_.addData(data);
+    state_ = State::Running;
+    coroutine_->resume(luaBody(coroutine_->luaState()), yield_callback_);
+  } else if (state_ == State::WaitForTrailers && end_stream_) {
+    ENVOY_LOG(debug, "resuming nil trailers due to end stream");
+    state_ = State::Running;
+    coroutine_->resume(0, yield_callback_);
+  }
+
+  if (state_ == State::HttpCall || state_ == State::WaitForBody) {
```

Comment:
It seems that we have to modify this global `State` for each new blocking feature we add.. can you comment on the tradeoffs here?